src/xdp/ebpfextension.c

```diff
@@ -0,0 +1,277 @@
+
+    ASSERT(Client != NULL);
+    _Analysis_assume_(Client != NULL);
+
+    NmrProviderDetachClientComplete(Client->NmrBindingHandle);
+
+    TraceExitSuccess(TRACE_CORE);
+}
+
+static
+NTSTATUS
+EbpfExtensionProviderAttachClient(
+    _In_ HANDLE NmrBindingHandle,
+    _In_ const VOID *ProviderContext,
+    _In_ const NPI_REGISTRATION_INSTANCE* ClientRegistrationInstance,
+    _In_ const VOID *ClientBindingContext,
+    _In_ const VOID *ClientNpiDispatch,
+    _Outptr_ VOID **ProviderBindingContext,
+    _Outptr_result_maybenull_ const VOID **ProviderDispatch
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    EBPF_EXTENSION_PROVIDER *Provider = (EBPF_EXTENSION_PROVIDER *)ProviderContext;
+    EBPF_EXTENSION_CLIENT *Client = NULL;
+    const ebpf_extension_dispatch_table_t *ClientDispatch;
+
+    TraceEnter(TRACE_CORE, "ProviderContext=%p", ProviderContext);
+
+    if ((ProviderBindingContext == NULL) || (ProviderDispatch == NULL) || (Provider == NULL)) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    *ProviderBindingContext = NULL;
+    *ProviderDispatch = NULL;
+
+    Client = ExAllocatePoolZero(NonPagedPoolNx, sizeof(*Client), XDP_POOLTAG_EBPF_NMR);
+    if (Client == NULL) {
+        Status = STATUS_NO_MEMORY;
+        goto Exit;
+    }
+
+    Client->NmrBindingHandle = NmrBindingHandle;
+    Client->ClientModuleId = ClientRegistrationInstance->ModuleId->Guid;
+    Client->ClientBindingContext = ClientBindingContext;
+    Client->ClientData =
+        (const ebpf_extension_data_t *)ClientRegistrationInstance->NpiSpecificCharacteristics;
+
+    ClientDispatch = (const ebpf_extension_dispatch_table_t *)ClientNpiDispatch;
+    if (ClientDispatch == NULL) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    Client->ClientDispatch = ClientDispatch;
+    Client->ProviderContext = Provider;
+
+    if (Provider->AttachCallback != NULL) {
+        Status = Provider->AttachCallback(Client, Provider);
+    }
+
+    if (NT_SUCCESS(Status)) {
+        Status = STATUS_SUCCESS;
+    } else {
+        Status = STATUS_ACCESS_DENIED;
+    }
+
+Exit:
+
+    if (NT_SUCCESS(Status)) {
+        *ProviderBindingContext = Client;
+        Client = NULL;
+    } else {
+        if (Client != NULL) {
+            ExFreePoolWithTag(Client, XDP_POOLTAG_EBPF_NMR);
+        }
+    }
+
+    TraceVerbose(
+        TRACE_CORE, "ProviderContext=%p ProviderBindingContext=%p",
+        ProviderContext, *ProviderBindingContext);
+    TraceExitStatus(TRACE_CORE);
+    return Status;
+}
+
+static
+NTSTATUS
+EbpfExtensionProviderDetachClient(
+    _In_ const VOID *ProviderBindingContext
+    )
+{
+    EBPF_EXTENSION_CLIENT *Client = (EBPF_EXTENSION_CLIENT *)ProviderBindingContext;
+    EBPF_EXTENSION_PROVIDER *Provider = Client->ProviderContext;
+    NTSTATUS Status = STATUS_SUCCESS;
+
+    TraceEnter(TRACE_CORE, "ProviderBindingContext=%p", ProviderBindingContext);
+
+    if (Client == NULL) {
```

Comment:
Is this something worth debug asserting?

Reply:
Hmm, yeah, we should assert. This check is copied from the netebpfext, which may have been needlessly paranoid - IMO we should be able to trust the NMR client to do the right thing here.